### PR TITLE
Issue 14582 (visitedDate not shown on UI)

### PR DESCRIPTION
**Pull request by baiti**
See description of the issue for more details

## Description

Extract visitedDate from the logs of the current user if the logType is either 2 (found it) or 10 (attended)

## Related issues

#14582

## Additional context

I am not sure  whether `getUserName()` gets me the uername from Opencaching.de or Geocaching.com. In my tests it worked, which may be due to tha fact that I am using the same username on OC and GC. Could somebody check if I am using the right approach for getting the username?

### CODE REVIEW

**Comment by fm-sys** [comment]
`Settings.getUserName()` can't know which connector you are requesting so it's most likely GC. 

OT:
IMHO we should deprecate the getUserName() method totally as it doesn't fit to the connector based concept of cgeo.

**Comment by baiti** [comment]
Slightly OT ... but still somewhat related: I spent some time looking at the code and it seems we don't have a concept of managing a `userid` based on the connector. `Settings.getUserName()` seems to give me the GC user. What if somebody wants to exlusively use OC and not GC? What if someone maintains different usernames on GC and OC?

I think this is a bigger problem than just the "visitedDate" and further, I do think it should be resolved one way or another.

In the OC case: there is an OKAPI service called "userinfo". This returns the username, among other things. But it doesn't make any sense to call that service each and every time I want to parse a geocache's log for the `visitedDate`. It must be possible, to store that variable somewhere I can later access it without running a costly OKAPI method.

I have tried to find such a way but my Java skills are way way way not sufficient to accomplish such a task. I repeatedly ran into  compile errors because things were "static" where they shouldn't be static and the othe way around. The error messages I got just confused me more, the more I tried.

**Comment by moving-bits** [comment]
> Slightly OT ... but still somewhat related:

I've created a separate issue for this for further discussion.

File: main/src/main/java/cgeo/geocaching/connector/oc/OkapiClient.java [hunk 1]
```diff
@@ -809,9 +810,19 @@ private static Geocache parseCache(final ObjectNode response) {
             cache.setLogPasswordRequired(response.get(CACHE_REQ_PASSWORD).asBoolean());
 
             cache.setDetailedUpdatedNow();
+
+            final List<LogEntry> logs = parseLogs((ArrayNode) response.path(CACHE_LATEST_LOGS), cache.getGeocode());
+            final String me = getUserName();
```
Comment (moving-bits) [inline main/src/main/java/cgeo/geocaching/connector/oc/OkapiClient.java:815]: This won't work in cases where OKAPI username is different from GC username (or GC isn't used at all).

**Comment by eddiemuc** [comment]
Additional to the username discussion, the solution will only work if the current user's found log is among the latest logs on the cache. This is because we only retrieve the last x logs for a cache of course. (Don't know how many, I assume something around 40 or 50?)

**Comment by moving-bits** [comment]
Instead of using this workaround for guessing a user's found log we may stick to the approach used by c:geo for presenting a "found date" (although that may have drawbacks as well, but that's a different story), and investigate why that mechanism does not get triggered for OC-based caches. See https://github.com/cgeo/cgeo/issues/14582#issuecomment-1732253748

**Comment by baiti** [comment]
> Additional to the username discussion, the solution will only work if the current user's found log is among the latest logs on the cache. This is because we only retrieve the last x logs for a cache of course. (Don't know how many, I assume something around 40 or 50?)

As far as I remember the "fields" say to get them all. Otherwise I wouldn't have coded it that way.
PS: I was a bit surprised about the decision to ge them all but the rationale most likely was that there aren't many logs on any of thise OC caches anyway.

**Comment by baiti** [comment]
I gave it another try. It worked on my emulater. Perhaps the

`params.add("user_logs_only", "true");`

should be replaced with some sort of a constant? Ideas how it should be named then?

File: main/src/main/java/cgeo/geocaching/connector/oc/OkapiClient.java (hunk 1 above)
Comment (baiti) [reply main/src/main/java/cgeo/geocaching/connector/oc/OkapiClient.java]: pls re-review

**Comment by fm-sys** [comment]
Just came to my mind again: We already store the user name for the FoundNumCounter on a per-connector base...

**Review by moving-bits — APPROVED** [review]
LGTM, thanks for adapting the PR

**Pull request merged**